common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java

```diff
@@ -146,6 +158,102 @@ protected TokenResult performTokenRequest(final OAuth2Strategy strategy,
         return tokenResult;
     }
 
+    protected void renewAccessToken(@NonNull final AcquireTokenSilentOperationParameters parameters,
+                                  @NonNull final AcquireTokenResult acquireTokenSilentResult,
+                                  @NonNull final OAuth2TokenCache tokenCache,
+                                  @NonNull final OAuth2Strategy strategy,
+                                  @NonNull final ICacheRecord cacheRecord)
+            throws IOException, ClientException, UiRequiredException {
+        final String methodName = ":renewAccessToken";
+        Logger.verbose(
+                TAG + methodName,
+                "Renewing access token..."
+        );
+        parameters.setRefreshToken(cacheRecord.getRefreshToken());
+
+        final TokenResult tokenResult = performSilentTokenRequest(strategy, parameters);
+        acquireTokenSilentResult.setTokenResult(tokenResult);
+
+        if (tokenResult.getSuccess()) {
+            Logger.verbose(
+                    TAG + methodName,
+                    "Token request was successful"
+            );
+            final ICacheRecord savedRecord = tokenCache.save(
+                    strategy,
+                    getAuthorizationRequest(strategy, parameters),
+                    tokenResult.getTokenResponse()
+            );
+
+            // Create a new AuthenticationResult to hold the saved record
+            final LocalAuthenticationResult authenticationResult = new LocalAuthenticationResult(savedRecord);
+
+            // Set the AuthenticationResult on the final result object
+            acquireTokenSilentResult.setLocalAuthenticationResult(authenticationResult);
+        } else {
+            // Log all the particulars...
+            if (null != tokenResult.getErrorResponse()) {
```

Comment:
Should `L#193` be an `else if(null != tokenResult.getErrorResponse())`?

What code path would be go down if the server returned an empty response body? How would we detect this?

What if there were an error response that was **not** `invalid_grant`? Would we see it? Would we log it?